README.md
# Gfret
Contents
========
* [Introduction](#introduction)
* [Getting Gfret](#getting-gfret)
* [Usage](#usage)
  * [Cli](#running-the-command-line-interface)
  * [Gui](#running-the-gui)
    * [Keybindings](#keybindings)
  * [Templates](#templates)
* [Configuration](#configuration)
* [Building](#building)
* [Roadmap](#roadmap)

## Introduction
*Gfret* renders an svg image template of a fretboard for a stringed instrument.
It has a Gtk interface as well as a command line interface and can produce
templates for instruments ranging from a piccolo mandolin to an upright bass.
Multiscale designs are also supported. Currently, all measurements are
expressed in metric units only.
## Getting Gfret
*Gfret* is distributed primarily in source form. The main repo is at
[codeberg.org](https://codeberg.org/jeang3nie/gfret) with mirrors at
[gitlab](https://gitlab.com/jeang3nie/gfret) and
[gibub.com](https://github.com/nfisher1226/gfret). Releases are recommended.
## Usage
*Gfret* can be used from the command line or using the Gtk+ interface.
## Running the command line interface
```Bash
gfret-cli

Output an svg without running the interface

USAGE:
   gfret cli [OPTIONS] [OUTPUT]

ARGS:
   <OUTPUT>    Name of the output file [default: output.svg]

OPTIONS:
   -b, --bridge <BRIDGE>
           Bridge spacing [default: 56]

   -c, --count <COUNT>
           Total fret count [default: 24]

   -e, --external <EXTERN>
           Open output file in external program [default: inkscape]

   -h, --help
           Print help information

   -l, --left
           Multiscale fretboard reversed (left handed)

   -m, --multi <MULTI>
           Creates a multiscale fretboard with <MULTI> as the treble scale. [default: 610]

   -n, --nut <NUT>
           Nut width [default: 43]

   -o, --output <OUTPUT>
           Name of the output file [default: output.svg]

   -p, --perpendicular <PERPENDICULAR>
           Set which fret is perpendicular to the centerline [default: 8]
   -s, --scale <SCALE>
           Scale length in mm [default: 648]
```
## Running the gui
Calling the program by invoking ```gfret``` without any arguments will run
the Gtk interface. Additionally, a .desktop file and icon are included and
will be installed if the program is installed using the included
```Makefile```, and can be used for launching the program from desktop menus
or creating shortcuts.
## Keybindings
> Note: See [KEYS.md](KEYS.md) for information on changing keybindings

| Key | Action |
| --- | --- |
| Ctrl/S | save file |
| Ctrl/Shift/S | save file as |
| Ctrl/E | open with an external program |
| Ctrl/O | load a template from file |
| Ctrl/Shift/P | open the `preferences` dialog |
| Ctrl/A | open the `about` dialog |
| Ctrl/Q | quit the program |
## Templates
Along with the svg output, Gfret will save the specifications used to
generate the rendering in a Toml file with it's name corresponding to the
name of the svg file. These templates can be loaded later, either as an
argument when invoking the program, in which case the output will be
immediately generated, or else loaded from the Gui interface for further
editing. This is useful for sharing a common scale among multiple designs to
use as a starting point.
## Configuration
On Unix systems the default configuration directory is ```~/.config/gfret```.
Gfret will maintain a configuration file here in [Toml](https://github.com/toml-lang/toml)
format, with the following fields:
```Toml
## must match either "Metric" or "Imperial"
units = "Metric"
## the command line to run
external_program = "inkscape"
## the size of the border around the image
border = 10.0
## how thick the lines are
line_weight = 1.0

[fretline_color]
## Colors must be "Hex", "Reduced" or "RGBA"
## "Reduced" will take a whole number between 0 and 255 for each channel,
## while "RGBA" takes a decimal between 0.0 and 1.0.

## "Hex" will have a <color> field and an <alpha> field
## ColorType = "Hex"
## color = "#00ff00"
## alpha = 1.0
ColorType = "Reduced"
red = <u8>
green = <u8>
blue = <u8>
alpha = <u8>

[fretboard_color]
ColorType = "Reduced"
red = 36
green = 31
blue = 49
alpha = 255

[centerline_color]
ColorType = "Reduced"
red = 0
green = 0
blue = 255
alpha = 255

## Fonts take a family and weight (style), but size is ignored
[font]
family = "Sans"
weight = "Normal"
```
> Note: The graphical interface has a preferences dialog and will take care
> of maintaining the preferences file for you. There will be no need to edit
> this file by hand in normal use.
## Building
You will need a Rust toolchain installed, including cargo. Gtk+4x is also
required.
```sh
cargo build --release
```
If desired, a release distribution can then be generated which will include the
binary, svg and png icons, Unix man pages, XDG .desktop file, gschema.xml and
shell completions.
```sh
cargo xtask dist
```
To install, copy the contents of `target/dist` to the appropriate prefix and
compile the gschemas.
```Sh
cp -Rv target/dist/* /usr/local
glib-compile-schemas /usr/local/share/glib-2.0/gschemas/
```
## Roadmap
- [x] For the gui, it would be nice to save state and allow loading specs
  from and saving to templates. **partial implementation 4/7/21** |
  **completed 5/5/21**
- [x] Port to Gtk4 **completed 12/21**
- [x] Support changing from metric to imperial measurements **completed 12/21**
- [x] Support left handed multiscale fretboards **completed 12/21**
- [x] Orient left handed output with bass strings on top **completed 1/22**
### 3.0
- [x] Subclass application window from AdwApplicationWindow
- [x] Subclass GfretApplication from AdwApplication
- [ ] Create a gio::Settings object and store application state in it
- [ ] Move Config into GfretApplication
- [x] Provide theme switcher
- [x] Subclass preferences window from AdwPreferencesWindow
- [ ] Make a property action and menu entry for Units
- [ ] Add AdwToastOverlay and set toast when a file is saved
- [ ] Replace gfret::File type for glib::File
- [ ] Make a `changed` boolean property of GfretWindow to track saved state
- [ ] Create all dialogs as needed instead of persisting for the program lifespan
- [ ] Replace templates with the ability to re-open svg files created by this
      application


RELEASE_NOTES.md
Contents
========
* [unreleased](#unreleased)
* [2.1](#2.1-release)
* [2.0.2](#2.0.2-release)
* [2.0.1](#2.0.1-release)
* [2.0](#2.0-release)

## Unreleased
* Move some common code into lib.rs
* Create trait `ConvertUnits` to swap imperial and metric values
* Move action handling into module
* Move keybindings into module
* Make keybindings configurable
* Depend on libadwaita
* Subclass Application from AdwApplication
* Subclass Window from AdwWindow
* Use adwaita AboutWindow
* Create Application::settings object and store application state

## 2.1 release
* Update to `gtk4-rs` version `0.4`
* Orient left handed output with bass strings on top (via `fretboard-layout` crate)
* Update `clap` to 3.0 release branch

## 2.0.2 release
* Preferences dialog updated around external application
* The external application now remains consistent the next time the dialog is
  opened
* Spinbutton fields now change precision to track better against Metric vs
  Imperial units

## 2.0.1 release
* Update `fretboard_layout` crate with cleaned up backend code
* Less file IO due to keeping the Config data in a static Mutex
* Where appropriate, the spinbox adjustments now display 2 digits for Metric
  and 3 digits when using Imperial units

## 2.0 release
This is the second major release of *gfret*, bringing a major code cleanup and
reorganization and several new features. The big news is that the interface has
been ported to [Gtk+ version 4](https://drewdevault.com/blog/index.xml). This
necessitated a number of changes to match syntax as well as some rethinking of
the basic assumptions around which the interface had previously been built.

The [fretboard-layout](https://crates.io/crates/fretboard_layout) crate which
serves as the backend to *gfret* has seen significant improvements as well. The
logic which handled anything color related has been moved into a new crate called
[rgba-simple](https://crates.io/crates/rgba_simple). Two new abilities have been
added, the ability to output right or left handed output, and the ability to
output Imperial measurements as well as Metric.
* Port interface to `Gtk4`
  * Port `.glade` files to new `.ui` interface definitions
  * Dialogs moved into a separate module
  * Preferences dialog significantly cleaned up with more consistent padding
    and spacing between elements
  * Dialogs all now use a `headerbar`
  * Main window changed to a `headerbar` and the application menu moved into it
  * Menu ported to new menu definition framework
  * Menu entries and keyboard shortcuts now use the `gtk::SimpleAction` construct
* Enable left handed output
  * Gui changed from a simple `checkbox` enabling the multiscale output to a
    `combobox`.
  * When `Monoscale` is selected, all `Multiscale` related controls are
    completely hidden rather than just greyed out.
  * Internally, the `Specs` struct represents whether the neck is Mono or Multi
    scaled using the `Variant` enum, which has an attached scale length and
    `Handedness` enum when `Multiscale` is selected.
* Enable switching between `Imperial` or `Metric` units
  * The `Config` struct has a new field `units` which is represented by an enum
    `Units`
  * All widget values *and* ranges are adjusted by a factor of 20.4 when
    switching between units
* Revamped `Preferences` dialog
  * Consistent padding and spacing
  * Controls which are no longer valid in the current context are completely
    hidden from the user rather than just being greyed out as before
* New `About` dialog
* New application icon with fretboard and dividers


src/cli.rs
#![warn(clippy::all, clippy::pedantic)]
use {
    clap::{Arg, ArgMatches, Command},
    fretboard_layout::{Handedness, Specs, Variant},
    std::process,
};

pub fn build_cli() -> Command<'static> {
    Command::new("cli")
        .about("Output an svg without running the interface")
        .long_about(
            "Gfret can be run from the commmand line just as easily as it can from a graphical
interface, and sometimes more quickly if all of the desired specifications are
known in advance.",
        )
        .author("The JeanG3nie <jeang3nie@hitchhiker-linux.org>")
        .version(env!("CARGO_PKG_VERSION"))
        .arg(
            Arg::new("SCALE")
                .help("Scale length in mm")
                .short('s')
                .long("scale")
                .takes_value(true)
                .default_value("648"),
        )
        .arg(
            Arg::new("MULTI")
                .help("Multiscale fretboard with <MULTI> as the treble scale")
                .short('m')
                .long("multi")
                .takes_value(true)
                .default_value("610"),
        )
        .arg(
            Arg::new("LEFT")
                .help("Multiscale fretboard reversed (left handed)")
                .short('l')
                .long("left")
                .takes_value(false),
        )
        .arg(
            Arg::new("PERPENDICULAR")
                .help("Which fret is perpendicular to the centerline")
                .short('p')
                .long("perpendicular")
                .takes_value(true)
                .default_value("8"),
        )
        .arg(
            Arg::new("COUNT")
                .help("Total fret count")
                .short('c')
                .long("count")
                .takes_value(true)
                .default_value("24"),
        )
        .arg(
            Arg::new("NUT")
                .help("Nut width")
                .short('n')
                .long("nut")
                .takes_value(true)
                .default_value("43"),
        )
        .arg(
            Arg::new("BRIDGE")
                .help("Bridge Spacing")
                .short('b')
                .long("bridge")
                .takes_value(true)
                .default_value("56"),
        )
        .arg(
            Arg::new("OUTPUT")
                .help("Name of the output file")
                .takes_value(true)
                .default_value("output.svg"),
        )
        .arg(
            Arg::new("EXTERN")
                .help("Open output file in external program")
                .short('e')
                .long("external")
                .takes_value(true)
                .default_value("inkscape"),
        )
}

pub fn build() -> Command<'static> {
    Command::new("gfret")
        .about("Generates layout dimensions for a stringed instrument fretboard")
        .author("The JeanG3nie <jeang3nie@hitchhiker-linux.org>")
        .version(env!("CARGO_PKG_VERSION"))
        .arg(
            Arg::new("TEMPLATE")
                .help("A valid Gfret template in toml format")
                .short('t')
                .long("template")
                .takes_value(true),
        )
        .subcommand(build_cli())
}

/// Runs the cli
pub fn run(matches: &ArgMatches) {
    let scale: f64 = match matches.value_of_t("SCALE") {
        Ok(c) => c,
        Err(e) => {
            eprintln!("{}", e);
            process::exit(1);
        }
    };
    let count = match matches.value_of_t("COUNT") {
        Ok(c) => c,
        Err(e) => {
            eprintln!("{}", e);
            process::exit(1);
        }
    };
    let variant = if matches.occurrences_of("MULTI") > 0 {
        let scale = match matches.value_of_t("MULTI") {
            Ok(c) => c,
            Err(e) => {
                eprintln!("{}", e);
                process::exit(1);
            }
        };
        let hand = if matches.occurrences_of("LEFT") > 0 {
            Handedness::Left
        } else {
            Handedness::Right
        };
        Variant::Multiscale(scale, hand)
    } else {
        Variant::Monoscale
    };
    let nut = match matches.value_of_t("NUT") {
        Ok(c) => c,
        Err(e) => {
            eprintln!("{}", e);
            process::exit(1);
        }
    };
    let bridge = match matches.value_of_t::<f64>("BRIDGE") {
        Ok(c) => c + 6.0,
        Err(e) => {
            eprintln!("{}", e);
            process::exit(1);
        }
    };
    let pfret = match matches.value_of_t("PERPENDICULAR") {
        Ok(c) => c,
        Err(e) => {
            eprintln!("{}", e);
            process::exit(1);
        }
    };
    let specs = Specs::init(scale, count, variant, nut, bridge, pfret);
    let doc = specs.create_document(None);
    let output = matches.value_of("OUTPUT").unwrap().to_string();
    if output == "-" {
        println!("{}", doc);
    } else {
        match svg::save(&output, &doc) {
            Ok(_) => println!("Output saved as {}.", output),
            Err(e) => {
                eprintln!("{}", e);
                process::exit(1);
            }
        };
        if matches.occurrences_of("EXTERN") > 0 {
            if let Some(cmd) = matches.value_of("EXTERN") {
                match process::Command::new(&cmd).args(&[&output]).spawn() {
                    Ok(_) => (),
                    Err(e) => eprintln!("{}", e),
                }
            }
        }
    }
}
